File: Shorting/quicksort.c
#include "stdio.h"

void quickSort(int arr[], int low, int high);
int partition(int arr[], int low, int high);
void swap(int* a, int* b);

void quickSort(int arr[], int low, int high) {
    if(low < high) {
        int part = partition(arr, low, high);

        quickSort(arr, low, part - 1);
        quickSort(arr, part + 1, high);
    }
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int index = low - 1;

    for(int i = low; i <= high - 1; i++) {
        if(arr[i] < pivot) {
            index++;
            swap(&arr[index], &arr[i]);
        }
    }
    swap(&arr[index + 1], &arr[high]);
    return index + 1;
}

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int main(void) {
    int arr[] = {9, 8, 7, 6, 5};
    int arraySize = sizeof(arr) / sizeof(int);
    quickSort(arr, 0, arraySize - 1);
    printf("Sorted Array:\n");
    for(int i = 0; i < arraySize; i++) {
        printf("%d", arr[i]);
    }
}

File: Sorting/Java/QuickSort.java
import java.util.*;

public class QuickSort{
	public static void display(int arr[], int left, int right){
		int i;
		System.out.println("I: "+left+" J: "+right);
		for(i=0;i<arr.length;i++)	
				System.out.print(arr[i]+"\t");
		System.out.println("");
	}

	public static int partition(int arr[], int left, int right){
		int i=left, j=right, temp;

		int pivot = arr[(left+right)/2];
                System.out.println("Pivot: "+pivot);
                  
		while(i<=j){
			while(arr[i]<pivot)
				i++;
			while(arr[j]>pivot)
				j--;
			if(i<=j){
				temp = arr[i];
				arr[i] = arr[j];
				arr[j] = temp;
				i++;
				j--;
			}	
		}
		System.out.println("After Partition: ");
		display(arr, i, j);
		return i;
	}   // <pivot | pivot | >pivot
	
	public static void quickSort(int arr[], int left, int right){
		int index = partition(arr, left, right);
		if(left<index-1){ 
			System.out.println("When left > index-1 :"+left+" > "+(index-1));
			display(arr, left, index-1);
			quickSort(arr, left, index-1);
		}
		if(index<right){
			System.out.println("When right > index :"+right+" > "+index);
			display(arr, index, right);
			quickSort(arr, index, right);
		}
		//display(arr, left, right);
	}
	
	public static void main(String args[]){
		Scanner sc = new Scanner(System.in);
		//int n = sc.nextInt();
		int i;
		//int arr[] = {18, 11, 19, 98, 7, 10, 70};
		
                int arr[] = {38, 27, 42, 9};
                quickSort(arr, 0, arr.length-1);		
	}
}	
